File: src_c/perftree.c
#include "board.h"
#include "movegen.h"
#include "util.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* #define printf(...) \ */
/*   LOG(__VA_ARGS__); \ */
/*   printf(__VA_ARGS__) */

int process_moves(struct Board *b, char *moves) {
  char *token = strtok(moves, " ");
  while (token) {
    Move m = move_from_str(b, token);
    if (m == NULL_MOVE) {
      return 1;
    }

    copy_make(b, b, m);
    token = strtok(NULL, " ");
  }

  return 0;
}

long perftree(struct Board *board, int depth) {
  if (depth == 0) {
    return 1;
  }

  long mc = 0;

  struct MoveList ml = new_move_list();
  bool checked = is_in_check(board);
  gen_moves(board, &ml, checked);

  struct Board b;
  for (int i = 0; i < ml.count; i++) {
    Move m = ml.moves[i];
    copy_make(board, &b, m);

    if (!is_legal_move(&b, m, checked)) {
      continue;
    }

    mc += perftree(&b, depth - 1);
  }

  return mc;
}

void perftree_root(struct Board *board, int depth) {
  long total_mc = 0;

  struct MoveList ml = new_move_list();
  bool checked = is_in_check(board);
  gen_moves(board, &ml, checked);

  struct Board b;
  for (int i = 0; i < ml.count; i++) {
    Move m = ml.moves[i];
    copy_make(board, &b, m);

    char move_str[10];
    uci_from_move(move_str, m);

    if (!is_legal_move(&b, m, checked)) {
      LOG("Skipping:\n");
      LOGMOVE(m);
      LOGBOARD(board);
      LOGBOARD(&b);
      continue;
    }

    long mc = perftree(&b, depth - 1);
    LOG("%s %ld\n", move_str, mc);
    printf("%s %ld\n", move_str, mc);
    total_mc += mc;
  }

  LOG("\n%ld\n", total_mc);
  printf("\n%ld\n", total_mc);
  fflush(stdout);
}

int main(int argc, char **argv) {
  init();

  if (argc < 3) {
    fprintf(stderr, "Not enough argumets, need [fen] [depth]\n");
    return 1;
  }

  for (int i = 1; i < argc; i++) {
    LOG("%d: %s\n", i, argv[i]);
  }

  int depth = atoi(argv[1]);

  if (depth < 1) {
    fprintf(stderr, "Invalid depth: %s\n", argv[1]);
    return 1;
  }

  struct Board b;
  if (board_from_fen(&b, argv[2])) {
    fprintf(stderr, "Invalid fen: %s\n", argv[2]);
    return 1;
  }

  if (argc == 4) {
    if (process_moves(&b, argv[3])) {
      fprintf(stderr, "Could not process moves: %s\n", argv[3]);
      return 1;
    }
  }

  LOGBOARD(&b);
  LOGCASTLE(b.castling);
  perftree_root(&b, depth);
  return 0;
}

File: src/perftree.zig
const std = @import("std");
const log = std.log;
const board = @import("board.zig");
const Board = board.Board;
const movegen = @import("movegen.zig");
const uci = @import("uci.zig");

fn perftree(b: *Board, depth: usize) usize {
    if (depth == 0) {
        return 1;
    }

    var mc: usize = 0;

    var ml = movegen.MoveList.new(b);
    const checked = b.is_in_check();
    movegen.gen_moves(&ml, b, checked);

    var next: Board = undefined;
    while (ml.next()) |m| {
        b.copy_make(&next, m);

        if (!movegen.is_legal_move(&next, m, checked)) {
            continue;
        }

        mc += perftree(&next, depth - 1);
    }

    return mc;
}

fn perftree_root(w: anytype, b: *Board, depth: usize) !void {
    var total_mc: usize = 0;

    var ml = movegen.MoveList.new(b);
    const checked = b.is_in_check();
    movegen.gen_moves(&ml, b, checked);

    var next: Board = undefined;
    while (ml.next()) |m| {
        b.copy_make(&next, m);

        if (!movegen.is_legal_move(&next, m, checked)) {
            continue;
        }

        const mc = perftree(&next, depth - 1);
        total_mc += mc;
        try m.as_uci_str(w);
        try std.fmt.format(w, " {d}\n", .{mc});
    }

    try std.fmt.format(w, "\n{d}\n", .{total_mc});
}

pub fn main() !void {
    var it = std.process.args();
    _ = it.skip();

    var arg = it.next() orelse return error.NoDepthArg;
    const depth = try std.fmt.parseInt(usize, arg, 10);

    arg = it.next() orelse return error.NoFenArg;
    var b = try board.board_from_fen(arg);

    if (it.next()) |move_str| {
        b = try uci.process_moves(b, move_str);
    }

    log.debug("perftree position is:", .{});
    b.log(log.debug);

    if (it.skip()) {
        return error.TooManyArgs;
    }

    try perftree_root(std.io.getStdOut().writer(), &b, depth);
}
